src/kernel/src/main.rs

```diff
@@ -267,24 +267,43 @@ fn main() -> ExitCode {
 
 fn exec<E: ee::ExecutionEngine>(mut ee: E, arg: EntryArg) -> ExitCode {
     // TODO: Check how the PS4 allocate the stack.
-    // TODO: We should allocate a guard page to catch stack overflow.
     info!("Allocating application stack.");
 
-    let stack = match MemoryManager::current().mmap(
+    // Assign stack size with an extra page for guard
+    let stack_size = 0x200000;
+    let guard_size = MemoryManager::current().page_size();
+    // Combine both sizes into one
+    let total_size = stack_size + guard_size;
+
+    let mut stack = match MemoryManager::current().mmap(
         0,
-        0x200000,
+        total_size,
         arg.stack_prot(),
         MappingFlags::MAP_ANON | MappingFlags::MAP_PRIVATE,
         -1,
         0,
     ) {
         Ok(v) => v,
         Err(e) => {
-            error!(e, "Allocate failed");
+            error!(e, "Stack allocation failed");
             return ExitCode::FAILURE;
         }
     };
 
+    // Calculate the aligned address for the guard page
+    let guard_start = (stack.as_mut_ptr() as usize + guard_size - 1) & !(guard_size - 1);
```

Comment:
Why guard the page before the last page?

Reply:
Oh, my b. Seems I accidentally put it in the wrong spot x3

Reply:
All you need to do here is passing `stack.as_mut_ptr()` directly as a first argument of `mprotect`.